aptos-move/framework/aptos-stdlib/sources/cryptography/curves.move

```diff
@@ -0,0 +1,217 @@
+    public fun scalar_zero<G>(): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_zero_internal(get_group_id<G>())
+        }
+    }
+
+    public fun scalar_one<G>(): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_one_internal(get_group_id<G>())
+        }
+    }
+
+    public fun scalar_from_u64<G>(value: u64): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_from_u64_internal(value, get_group_id<G>())
+        }
+    }
+
+    public fun scalar_neg<T>(_scalar_1: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_neg_internal(get_group_id<T>())
+        }
+    }
+
+    public fun scalar_add<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_add_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_mul<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_mul_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_inv<T>(scalar: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_inv_internal(scalar.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_eq<T>(scalar_1: &Scalar<T>, scalar_2: &Scalar<T>): bool {
+        scalar_eq_internal(scalar_1.handle, scalar_2.handle, get_group_id<T>())
+    }
+
+    public fun bytes_into_scalar<T>(bytes: vector<u8>): Scalar<T> {
+        Scalar<T> {
+            handle: bytes_into_scalar_internal(bytes, get_group_id<T>())
+        }
+    }
+
+    // Point basics.
+    public fun point_identity<T>(): Point<T> {
+        Point<T> {
+            handle: point_identity_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_generator<T>(): Point<T> {
+        Point<T> {
+            handle: point_generator_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_add<T>(point_1: &Point<T>, point_2: &Point<T>): Point<T> {
+        Point<T> {
+            handle: point_add_internal(point_1.handle, point_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun point_mul<G>(_scalar: &Scalar<G>, _point: &Point<G>): Point<G> {
+        Point<G> {
+            handle: point_mul_internal(_scalar.handle, _point.handle, get_group_id<G>())
+        }
+    }
+
+    public fun scalar_to_bytes<T>(scalar: &Scalar<T>): vector<u8> {
+        scalar_to_bytes_internal(scalar.handle, get_group_id<T>())
+    }
+
+    public fun point_to_bytes<T>(point: &Point<T>): vector<u8> {
+        point_to_bytes_internal(point.handle, get_group_id<T>())
+    }
+
+    public fun bytes_into_point<T>(bytes: vector<u8>): Point<T> {
+        Point<T> {
+            handle: bytes_into_point_internal(bytes, get_group_id<T>())
+        }
+    }
+
+    public fun point_eq<T>(point_1: &Point<T>, point_2: &Point<T>): bool {
+        point_eq_internal(point_1.handle, point_2.handle, get_group_id<T>())
+    }
+
```

Comment:
We'll also want `random_point` and `random_scalar` #[test-only]` functions, which we can use in our tests.

Reply:
We'll also want a `point_is_prime_order<T>(point: &Point<T>): bool` method which returns true if the point is of prime order.

Reply:
Oh, and we'll need `point_clone` and `scalar_clone` functions.

Reply:
If the `Element`/`Scalar` struct has copy ability and it stores only a handle, i think we don't need `point_clone`/`scalar_clone`.

Reply:
I think we do. There's an invariant somewhere in the code that you cannot have two Move struct objects with the same handle $x$.

I think if we cloned via `copy` trait and one object with handle $x$ went out of scope and got cleaned, the other one would be a dangling reference now.

But I don't remember how things get cleaned when they go out of scope, if they even get cleaned...

Reply:
We need to look into this and (re)understand.

Reply:
Checking the implementation, we have a VM session created for every transaction. VM sessions also have full ownership of `NativeContextExtension`s. I think that means:
- In Rust, the EC points we generated will be safely stored as extension state during transaction execution (as long as we don't mutate it).
- In Move, the EC point handles will always be valid during transaction execution.
- Extension inner-states are cleaned up after the execution.

Maybe @vgao1996 can confirm?

Reply:
Right, so the reason I didn't have `copy` in `RistrettoPoint` is because `RistrettoPoint`'s are mutable. So if `a` were a `RistrettoPoint` and we did:
```
let x = a;
mul_assign(&mut a, &some_scalar);
```
Then this would also change `x` which would be unexpected.

But, in your case, your `Element`'s seem to be immutable. In that case, you can safely have the `copy` trait AFAICT.